dora/core/common/src/main/java/alluxio/membership/ServiceDiscoveryRecipe.java

```diff
@@ -0,0 +1,323 @@
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.membership;
+
+import alluxio.annotation.SuppressFBWarnings;
+import alluxio.exception.status.AlreadyExistsException;
+import alluxio.resource.LockResource;
+import alluxio.util.ThreadFactoryUtils;
+
+import com.google.common.base.Preconditions;
+import io.etcd.jetcd.ByteSequence;
+import io.etcd.jetcd.KeyValue;
+import io.etcd.jetcd.Txn;
+import io.etcd.jetcd.kv.TxnResponse;
+import io.etcd.jetcd.lease.LeaseKeepAliveResponse;
+import io.etcd.jetcd.op.Cmp;
+import io.etcd.jetcd.op.CmpTarget;
+import io.etcd.jetcd.op.Op;
+import io.etcd.jetcd.options.GetOption;
+import io.etcd.jetcd.options.PutOption;
+import io.etcd.jetcd.support.CloseableClient;
+import io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * ServiceDiscoveryRecipe for etcd, to track health status
+ * of all registered services.
+ */
+public class ServiceDiscoveryRecipe {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEtcdClient.class);
+  private static final String BASE_PATH = "/ServiceDiscovery";
+  AlluxioEtcdClient mAlluxioEtcdClient;
+  ScheduledExecutorService mExecutor;
+  String mClusterIdentifier = "";
+  // Will look like /ServiceDiscovery/<mClusterIdentifier>
+  String mRegisterPathPrefix = "";
+  @SuppressFBWarnings({"URF_UNREAD_FIELD"})
+  private final ReentrantLock mRegisterLock = new ReentrantLock();
+  final ConcurrentHashMap<String, ServiceEntity> mRegisteredServices = new ConcurrentHashMap<>();
+
+  /**
+   * CTOR for ServiceDiscoveryRecipe.
+   * @param client
+   * @param clusterIdentifier
+   */
+  public ServiceDiscoveryRecipe(AlluxioEtcdClient client, String clusterIdentifier) {
+    mAlluxioEtcdClient = client;
+    mAlluxioEtcdClient.connect();
```

Comment:
nit: swap these two lines so a failed client will not result in a partially-init object?

Reply:
done